hal/src/stm32f2xx/sleep_hal.cpp

```diff
@@ -100,6 +117,90 @@ static int configGpioWakeupSource(const hal_wakeup_source_base_t* wakeupSources)
+            }
+            NVIC_DisableIRQ(ADC_IRQn);
+            NVIC_ClearPendingIRQ(ADC_IRQn);
+
+            RCC_APB2PeriphClockCmd(RCC_APB2Periph_ADC1, ENABLE);
+            /* ADC Common Init */
+            ADC_CommonInitTypeDef ADC_CommonInitStructure = {};
+            ADC_CommonInitStructure.ADC_Mode = ADC_Mode_Independent;
+            ADC_CommonInitStructure.ADC_Prescaler = ADC_Prescaler_Div2;
+            ADC_CommonInitStructure.ADC_DMAAccessMode = ADC_DMAAccessMode_Disabled;
+            ADC_CommonInitStructure.ADC_TwoSamplingDelay = ADC_TwoSamplingDelay_5Cycles;
+            ADC_CommonInit(&ADC_CommonInitStructure);
+
+            // ADC1 configuration
+            ADC_InitTypeDef ADC_InitStructure = {};
+            ADC_InitStructure.ADC_Resolution = ADC_Resolution_12b; // 12-bits
+            ADC_InitStructure.ADC_ScanConvMode = DISABLE;
+            ADC_InitStructure.ADC_ContinuousConvMode = ENABLE;
+            ADC_InitStructure.ADC_ExternalTrigConvEdge = ADC_ExternalTrigConvEdge_None;
+            ADC_InitStructure.ADC_ExternalTrigConv = ADC_ExternalTrigConv_T1_CC1;
+            ADC_InitStructure.ADC_DataAlign = ADC_DataAlign_Right;
+            ADC_InitStructure.ADC_NbrOfConversion = 1;
+            ADC_Init(ADC1, &ADC_InitStructure);
+
+            // ADC1 regular channel configuration
+            ADC_RegularChannelConfig(ADC1, pinMap[analogWakeup->pin].adc_channel, 1, ADC_SampleTime_480Cycles);
+
+            // Analog watchdog configuration
+            ADC_AnalogWatchdogCmd(ADC1, ADC_AnalogWatchdog_SingleRegEnable);
+            ADC_AnalogWatchdogSingleChannelConfig(ADC1, pinMap[analogWakeup->pin].adc_channel);
+            uint16_t th = (uint16_t)(analogWakeup->voltage * (4095 / 3300.0));
+            th &= 0x00000FFF;
+            switch (analogWakeup->trig) {
+                case HAL_SLEEP_LPCOMP_ABOVE: {
+                    ADC_AnalogWatchdogThresholdsConfig(ADC1, th, 0);
+                    break;
+                }
+                case HAL_SLEEP_LPCOMP_BELOW: {
+                    ADC_AnalogWatchdogThresholdsConfig(ADC1, 4095, th);
+                    break;
+                }
+                case HAL_SLEEP_LPCOMP_CROSS: {
+                    if (currVol >= th) {
+                        ADC_AnalogWatchdogThresholdsConfig(ADC1, 4095, th);
+                    } else {
+                        ADC_AnalogWatchdogThresholdsConfig(ADC1, th, 0);
+                    }
+                    break;
+                }
+                default: return SYSTEM_ERROR_INVALID_ARGUMENT; // It shouldn't reach here.
+            }
+
+            // Interrupt configuration
+            ADC_ClearFlag(ADC1, ADC_FLAG_AWD);
+            ADC_ITConfig(ADC1, ADC_IT_AWD, ENABLE);
+
+            //select NVIC channel to configure
+            NVIC_InitTypeDef NVIC_InitStructure = {};
+            NVIC_InitStructure.NVIC_IRQChannel = ADC_IRQn;
+            NVIC_InitStructure.NVIC_IRQChannelPreemptionPriority = 7;
```

Comment:
This will probably also have the same problem as https://github.com/particle-iot/device-os/pull/2173#discussion_r475970347